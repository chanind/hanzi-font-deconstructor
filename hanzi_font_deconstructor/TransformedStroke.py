from dataclasses import dataclass
from typing import Tuple

@dataclass
class TransformedStroke:
    translate: Tuple[float, float]
    rotate: float
    skewX: float
    skewY: float
    scale: Tuple[float, float]
    path: str