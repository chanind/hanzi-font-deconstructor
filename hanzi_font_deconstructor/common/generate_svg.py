from dataclasses import dataclass
from .TransformedStroke import TransformedStroke
from typing import Sequence, Tuple


@dataclass
class StrokeAttrs:
    d: str
    transform: str


def generate_svg(
    strokes_attrs: Sequence[StrokeAttrs], viewbox=Tuple[int, int, int, int]
) -> str:
    svg_paths = []
    for attrs in strokes_attrs:
        svg_paths.append(f'<path d="{attrs.d}" transform="{attrs.transform}" />')
    viewboxStr = f"{viewbox[0]} {viewbox[1]} {viewbox[2]} {viewbox[3]}"
    svgPathsStr = "\n".join(svg_paths)
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="{viewboxStr}">\n'
        f"{svgPathsStr}\n"
        "</svg>"
    )


def get_stroke_attrs(stroke: TransformedStroke) -> StrokeAttrs:
    return StrokeAttrs(
        d=stroke.path,
        transform=(
            " ".join(
                [
                    f"translate({stroke.translate[0]}, {stroke.translate[1]})",
                    f"rotate({stroke.rotate})",
                    f"skewX({stroke.skewX})",
                    f"skewY({stroke.skewY})",
                    f"scale({stroke.scale[0]}, {stroke.scale[1]})",
                ]
            )
        ),
    )
