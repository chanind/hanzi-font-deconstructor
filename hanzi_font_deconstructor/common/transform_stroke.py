from svgpathtools import parse_path
from typing import Tuple
from .TransformedStroke import TransformedStroke
from random import randint, uniform, gauss


def transform_stroke(
    stroke_pathstr: str, viewbox: Tuple[int, int, int, int]
) -> TransformedStroke:
    path = parse_path(stroke_pathstr)
    strokeMinX, strokeMaxX, strokeMinY, strokeMaxY = path.bbox()
    vbMinX, vbMinY, vbWidth, vbHeight = viewbox
    vbMaxX = vbMinX + vbWidth
    vbMaxY = vbMinY + vbHeight
    strokeMidX = (strokeMaxX + strokeMinX) / 2
    strokeMidY = (strokeMaxY + strokeMinY) / 2

    rotate = uniform(-10, 10)
    skewX = uniform(-10, 10)
    skewY = uniform(-10, 10)
    scaleX = max(0.1, gauss(0.3, 0.2))
    scaleY = scaleX * uniform(0.9, 1.1)

    # from https://stackoverflow.com/a/11671373
    baseTranslateX = (1 - scaleX) * strokeMidX
    baseTranslateY = (1 - scaleY) * strokeMidY
    translateX = baseTranslateX + randint(
        int(vbMinX - strokeMinX), int(vbMaxX - strokeMaxX)
    )
    translateY = baseTranslateY + randint(
        int(vbMinY - strokeMinY), int(vbMaxY - strokeMaxY)
    )

    return TransformedStroke(
        translate=(translateX, translateY),
        rotate=rotate,
        scale=(scaleX, scaleY),
        skewX=skewX,
        skewY=skewY,
        path=stroke_pathstr,
    )
