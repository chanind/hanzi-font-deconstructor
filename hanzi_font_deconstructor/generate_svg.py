from .TransformedStroke import TransformedStroke
from typing import Sequence, Tuple


def generate_svg(stroke_paths: Sequence[TransformedStroke], viewbox=Tuple[int, int, int, int]) -> str:
    svg_paths=[]
    for stroke_path in stroke_paths:
        svg_paths.append((
            f'<path fill="currentColor" d="{stroke_path.path}"'
            'transform="'
            f'  rotate({stroke_path.rotate})'
            f'  translate({stroke_path.translate[0]}, {stroke_path.translate[0]})'
            f'  skewX({stroke_path.skewX})'
            f'  skewY({stroke_path.skewY})'
            '" />'
        ))
    viewboxStr = f"{viewbox[0]} {viewbox[1]} {viewbox[2]} {viewbox[3]}"
    svgPathsStr = "\n".join(svg_paths)
    return (
        '<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">\n'
        f'<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="{viewboxStr}">\n'
        f'{svgPathsStr}\n'
        '</svg>'
    )