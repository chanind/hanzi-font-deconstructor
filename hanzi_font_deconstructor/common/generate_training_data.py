from .generate_svg import generate_svg, get_stroke_attrs
from .transform_stroke import transform_stroke
from .transform_stroke import transform_stroke
from .svg_to_pil import svg_to_pil
from os import path
from pathlib import Path
import random
import re
import torch
from torchvision import transforms


PROJECT_ROOT = Path(__file__).parents[2]
GLYPH_SVGS_DIR = PROJECT_ROOT / "noto_glyphs"

# https://en.wikipedia.org/wiki/Stroke_(CJK_character)
SINGLE_STROKE_CHARS = [
    "一",
    "乙",
    "丨",
    "丶",
    "丿",
    "乀",
    "乁",
    "乚",
    "乛",
    "亅",
    "𠃊",
    "𠃋",
    "𠃌",
    "𠃍",
    "𠃑",
    "𠄌",
    "㇐",
    "㇀",
    "㇖",
    "㇇",
    "㇕",
    "㇆",
    "㇊",
    "㇅",
    "㇍",
    "㇈",
    "㇠",
    "㇎",
    "㇋",
    "㇌",
    "㇡",
    "㇑",
    "㇚",
    "㇙",
    "㇗",
    "㇄",
    "㇘",
    "㇟",
    "㇞",
    "㇉",
    "㇒",
    "㇓",
    "㇢",
    "㇜",
    "㇛",
    "㇔",
    "㇏",
    "㇝",
    "㇂",
    "㇃",
    "㇁",
]


STROKE_VIEW_BOX = (-10, 0, 1010, 1000)
MISC_SINGLE_STROKE_PATHS = [
    "M884 65l34 62c-131 40 -349 62 -523 72c-2 -18 -10 -44 -18 -61c173 -12 387 -36 507 -73z",
    "M542 409 l-60 26c-14 -47 -46 -122 -74 -178l57 -22c30 56 63 127 77 174z",
    "M703 378l-64 21c-8 -47 -32 -118 -53 -173l61 -18c23 54 47 124 56 170z",
    "M849 191l65 19c-32 75 -76 163 -113 220c-13 -9 -43 -23 -60 -30c40 -56 81 -137 108 -209z",
    "M253 417v359c21 9 43 27 78 46c63 34 150 40 258 40c113 0 269 -7 370 -19c-10 22 -22 60 -24 83 c-78 5 -248 10 -349 10c-116 0 -202 -10 -270 -46c-41 -22 -73 -50 -95 -50c-34 0 -82 52 -129 108l-53 -71c48 -46 99 -84 142 -100v-290h-128v-70h200z",
    "M267 239l-62 45c-27 -45 -88 -111 -142 -158l58 -39c53 43 117 108 146 152z",
    "M268 753l113 -80c6 22 18 51 26 66c-186 137 -214 160 -228 178c-8 -18 -28 -51 -43 -66c21 -14 58 -48 58 -92v-331h-145v-72h219v397z",
]


def get_file_for_char(char):
    code = hex(ord(char))[2:]
    return path.join(GLYPH_SVGS_DIR, f"{code}.svg")


SINGLE_STROKE_CHAR_PATHS = []
path_extractor = re.compile(r'\bd="([^"]+)"')
for char in SINGLE_STROKE_CHARS:
    char_file = get_file_for_char(char)
    with open(char_file, "r") as contents:
        char_svg = contents.read().replace("\n", "")
    path_match = path_extractor.search(char_svg)
    if not path_match:
        raise Exception(f"No SVG path found in char svg: {char}")
    SINGLE_STROKE_CHAR_PATHS.append(path_match[1])

SINGLE_STROKE_PATHS = MISC_SINGLE_STROKE_PATHS + SINGLE_STROKE_CHAR_PATHS

tensorify = transforms.ToTensor()


def img_to_greyscale_tensor(img):
    return tensorify(img)[3, :, :]


def get_mask_span(mask):
    "return a tuple of (horizontal span, vertical span)"
    horiz_max_vals = torch.max(mask, 0).values
    vert_max_vals = torch.max(mask, 1).values
    min_x = torch.argmax(horiz_max_vals).item()
    max_x = len(horiz_max_vals) - torch.argmax(torch.flip(horiz_max_vals, [0])).item()
    min_y = torch.argmax(vert_max_vals).item()
    max_y = len(vert_max_vals) - torch.argmax(torch.flip(vert_max_vals, [0])).item()
    return (max_x - min_x, max_y - min_y)


def is_stroke_good(mask, existing_masks) -> bool:
    # if this is the first stroke, then anything is fine
    if len(existing_masks) == 0:
        return True

    # TODO: this is probably really slow, might need to speed this up somehow
    mask_size = torch.sum(mask).item()
    # mask_span = get_mask_span(mask)
    for existing_mask in existing_masks:
        existing_mask_size = torch.sum(existing_mask).item()
        overlaps = torch.where(existing_mask + mask >= 2, 1, 0)
        overlaps_size = torch.sum(overlaps).item()
        if overlaps_size == 0:
            # if this is the second stroke, ensure there's an overlap
            # we should ensure there's at least 1 overlap per training sample
            return len(existing_masks) > 1
        # if the overlap is a large amount of either stroke, this is a bad stroke
        if overlaps_size / existing_mask_size > 0.25:
            return False
        if overlaps_size / mask_size > 0.25:
            return False
        # overlaps_span = get_mask_span(overlaps)
        # existing_mask_span = get_mask_span(existing_mask)
        # # if the overlap is a large amount of the span of either stroke, this is a bad stroke
        # if max(overlaps_span) / max(existing_mask_span) > 0.4:
        #     return False
        # if max(overlaps_span) / max(mask_span) > 0.4:
        #     return False
    return True


MASK_THRESHOLD = 0.3


def get_training_input_svg_and_masks(size_px):
    num_strokes = random.randint(3, 4)
    with torch.no_grad():
        strokes_attrs = []
        stroke_masks = []
        while len(strokes_attrs) < num_strokes:
            pathstr = random.choice(SINGLE_STROKE_PATHS)
            stroke = transform_stroke(pathstr, STROKE_VIEW_BOX)
            stroke_attrs = get_stroke_attrs(stroke)
            stroke_svg = generate_svg([stroke_attrs], STROKE_VIEW_BOX)
            stroke_img = svg_to_pil(stroke_svg, size_px, size_px)
            stroke_tensor = img_to_greyscale_tensor(stroke_img)
            stroke_mask = torch.where(stroke_tensor > MASK_THRESHOLD, 1, 0)

            if is_stroke_good(stroke_mask, stroke_masks):
                strokes_attrs.append(stroke_attrs)
                stroke_masks.append(stroke_mask)
        input_svg = generate_svg(strokes_attrs, STROKE_VIEW_BOX)
    return (input_svg, stroke_masks)


def get_training_input_and_mask_tensors(size_px=256):
    with torch.no_grad():
        input_svg, stroke_masks = get_training_input_svg_and_masks(size_px)

        input_img = svg_to_pil(input_svg, size_px, size_px)
        input_tensor = img_to_greyscale_tensor(input_img)
        mask_sums = torch.zeros(input_tensor.shape, dtype=torch.long)
        for stroke_mask in stroke_masks:
            mask_sums += stroke_mask

        # collapse all overlaps of more than 2 items into a single "overlap" class
        mask = torch.where(mask_sums > 2, 2, mask_sums)
        return (input_tensor.unsqueeze(0), mask)
