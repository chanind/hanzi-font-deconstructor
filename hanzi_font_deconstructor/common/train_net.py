import logging
import os
from tqdm import tqdm
import torch
import torch.nn as nn
from torch import optim
from .generate_training_data import RandomStrokesDataset
from .eval_net import eval_net

from torch.utils.tensorboard import SummaryWriter
from torch.utils.data import DataLoader


def train_net(
    net,
    device,
    total_samples=10000,
    max_strokes=5,
    size_px=512,
    epochs=5,
    batch_size=1,
    lr=0.001,
    val_portion=0.1,
    save_cp_dir=None,
    img_scale=1,
):

    n_val = int(total_samples * val_portion)
    n_train = total_samples - n_val
    train_loader = DataLoader(
        RandomStrokesDataset(n_train, max_strokes=max_strokes, size_px=size_px),
        batch_size=batch_size,
        num_workers=2,
    )
    val_loader = DataLoader(
        RandomStrokesDataset(n_val, max_strokes=max_strokes, size_px=size_px),
        batch_size=batch_size,
        num_workers=2,
    )

    writer = SummaryWriter(comment=f"LR_{lr}_BS_{batch_size}_SCALE_{img_scale}")
    global_step = 0

    logging.info(
        f"""Starting training:
        Epochs:          {epochs}
        Batch size:      {batch_size}
        Learning rate:   {lr}
        Training size:   {n_train}
        Validation size: {n_val}
        Device:          {device.type}
        Images scaling:  {img_scale}
    """
    )

    optimizer = optim.RMSprop(net.parameters(), lr=lr, weight_decay=1e-8, momentum=0.9)
    scheduler = optim.lr_scheduler.ReduceLROnPlateau(
        optimizer, "min" if net.n_classes > 1 else "max", patience=2
    )
    if net.n_classes > 1:
        criterion = nn.CrossEntropyLoss()
    else:
        criterion = nn.BCEWithLogitsLoss()

    for epoch in range(epochs):
        net.train()

        epoch_loss = 0
        with tqdm(
            total=n_train, desc=f"Epoch {epoch + 1}/{epochs}", unit="img"
        ) as pbar:
            for batch in train_loader:
                imgs = batch["image"]
                true_masks = batch["mask"]
                assert imgs.shape[1] == net.n_channels, (
                    f"Network has been defined with {net.n_channels} input channels, "
                    f"but loaded images have {imgs.shape[1]} channels. Please check that "
                    "the images are loaded correctly."
                )

                imgs = imgs.to(device=device, dtype=torch.float32)
                mask_type = torch.float32 if net.n_classes == 1 else torch.long
                true_masks = true_masks.to(device=device, dtype=mask_type)

                masks_pred = net(imgs)
                loss = criterion(masks_pred, true_masks)
                epoch_loss += loss.item()
                writer.add_scalar("Loss/train", loss.item(), global_step)

                pbar.set_postfix(**{"loss (batch)": loss.item()})

                optimizer.zero_grad()
                loss.backward()
                nn.utils.clip_grad_value_(net.parameters(), 0.1)
                optimizer.step()

                pbar.update(imgs.shape[0])
                global_step += 1
                if global_step % (n_train // (10 * batch_size)) == 0:
                    for tag, value in net.named_parameters():
                        tag = tag.replace(".", "/")
                        writer.add_histogram(
                            "weights/" + tag, value.data.cpu().numpy(), global_step
                        )
                        writer.add_histogram(
                            "grads/" + tag, value.grad.data.cpu().numpy(), global_step
                        )
                    val_score = eval_net(net, val_loader, device)
                    scheduler.step(val_score)
                    writer.add_scalar(
                        "learning_rate", optimizer.param_groups[0]["lr"], global_step
                    )

                    if net.n_classes > 1:
                        logging.info("Validation cross entropy: {}".format(val_score))
                        writer.add_scalar("Loss/test", val_score, global_step)
                    else:
                        logging.info("Validation Dice Coeff: {}".format(val_score))
                        writer.add_scalar("Dice/test", val_score, global_step)

                    writer.add_images("images", imgs, global_step)
                    if net.n_classes == 1:
                        writer.add_images("masks/true", true_masks, global_step)
                        writer.add_images(
                            "masks/pred", torch.sigmoid(masks_pred) > 0.5, global_step
                        )

        if save_cp_dir:
            try:
                os.mkdir(save_cp_dir)
                logging.info("Created checkpoint directory")
            except OSError:
                pass
            torch.save(
                net.state_dict(), os.path.join(save_cp_dir, f"CP_epoch{epoch + 1}.pth")
            )
            logging.info(f"Checkpoint {epoch + 1} saved !")

    writer.close()
