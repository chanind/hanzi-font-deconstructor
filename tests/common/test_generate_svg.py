from hanzi_font_deconstructor.common.TransformedStroke import TransformedStroke
from hanzi_font_deconstructor.common.generate_svg import generate_svg, get_stroke_attrs

def test_generate_svg(snapshot):
    strokes = [
        TransformedStroke(
            translate=(12,134),
            rotate=-3,
            skewX=3,
            skewY=-1,
            scale=(0.3, 1.7),
            path="M703 378l-64 21c-8 -47 -32 -118 -53 -173l61 -18c23 54 47 124 56 170z",
        ),
        TransformedStroke(
            translate=(-2,232),
            rotate=-3,
            skewX=-12,
            skewY=0,
            scale=(1, 0.9),
            path="M849 191l65 19c-32 75 -76 163 -113 220c-13 -9 -43 -23 -60 -30c40 -56 81 -137 108 -209z",
        )
    ]
    strokes_attrs = [get_stroke_attrs(stroke) for stroke in strokes]
    assert generate_svg(strokes_attrs, (-10, 0, 1010, 1000)) == snapshot