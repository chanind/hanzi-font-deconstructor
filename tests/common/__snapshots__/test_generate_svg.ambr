# name: test_generate_svg
  '
  <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="-10 0 1010 1000">
  <path d="M703 378l-64 21c-8 -47 -32 -118 -53 -173l61 -18c23 54 47 124 56 170z" transform="translate(12, 134) rotate(-3) skewX(3) skewY(-1) scale(0.3, 1.7)" />
  <path d="M849 191l65 19c-32 75 -76 163 -113 220c-13 -9 -43 -23 -60 -30c40 -56 81 -137 108 -209z" transform="translate(-2, 232) rotate(-3) skewX(-12) skewY(0) scale(1, 0.9)" />
  </svg>
  '
---
